core/src/main/java/hivemall/tools/map/MapRouletteUDF.java

```diff
@@ -0,0 +1,209 @@
+        }
+
+        this.mapOI = HiveUtils.asMapOI(argOIs[0]);
+        this.valueOI = HiveUtils.asNumberOI(mapOI.getMapValueObjectInspector());
+
+        if (argOIs.length == 2) {
+            ObjectInspector argOI1 = argOIs[1];
+            if (HiveUtils.isIntegerOI(argOI1) == false) {
+                throw new UDFArgumentException(
+                    "The second argument of map_roulette must be integer type: "
+                            + argOI1.getTypeName());
+            }
+            if (ObjectInspectorUtils.isConstantObjectInspector(argOI1)) {
+                long seed = HiveUtils.getAsConstLong(argOI1);
+                this._rand = new Random(seed); // fixed seed
+            } else {
+                this.seedOI = HiveUtils.asLongCompatibleOI(argOI1);
+            }
+        } else {
+            this._rand = new Random(); // random seed
+        }
+
+        return mapOI.getMapKeyObjectInspector();
+    }
+
+    @Override
+    public Object evaluate(DeferredObject[] arguments) throws HiveException {
+        Random rand = _rand;
+        if (rand == null) {
+            Object arg1 = arguments[1].get();
+            if (arg1 == null) {
+                rand = new Random();
+            } else {
+                long seed = HiveUtils.getLong(arg1, seedOI);
+                rand = new Random(seed);
+            }
+        }
+
+        Map<Object, Double> input = getObjectDoubleMap(arguments[0], mapOI, valueOI);
+        if (input == null) {
+            return null;
+        }
+
+        return rouletteWheelSelection(input, rand);
+    }
+
+    @Nullable
+    private static Map<Object, Double> getObjectDoubleMap(@Nonnull final DeferredObject argument,
+            @Nonnull final MapObjectInspector mapOI,
+            @Nonnull final PrimitiveObjectInspector valueOI) throws HiveException {
+        final Map<?, ?> m = mapOI.getMap(argument.get());
+        if (m == null) {
+            return null;
+        }
+        final int size = m.size();
+        if (size == 0) {
+            return null;
+        }
+
+        final Map<Object, Double> result = new HashMap<>(size);
+        double sum = 0.d;
+        int cnt = 0;
+        for (Map.Entry<?, ?> entry : m.entrySet()) {
+            Object value = entry.getValue();
+            if (value == null) {
+                continue;
+            }
+            final double v = PrimitiveObjectInspectorUtils.convertPrimitiveToDouble(value, valueOI);
+            if (v < 0) {
+                throw new UDFArgumentException(
+                    "Map value must be greather than or equals to zero: " + entry.getValue());
+            }
+
+            Object key = entry.getKey();
+            result.put(key, Double.valueOf(v));
+
+            sum += v;
+            cnt++;
+        }
+
+        if (result.size() < m.size()) {
+            // fillna with the avg value
+            final Double avg = Double.valueOf(sum / cnt);
```

Comment:
Yes，`avg /= m.size();` was revised, so Jerry in testcase ' testRoulette()' should be elected for the 3rd place